cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -695,6 +705,45 @@ void deleteApp(Id.Application appId, ApplicationSpecification spec) throws Excep
     }
   }
 
+  /**
+   * Delete the business metadata for the application and the programs.
+   */
+  private void deleteAppBusinessMetadata(Id.Application appId, ApplicationSpecification appSpec) {
+    // Remove metadata for the Application itself.
+    businessMds.removeMetadata(appId);
+
+    // Remove business metadata for the programs of the Application
+    // TODO: Need to remove this we support prefix search of metadata type.
+    // See https://issues.cask.co/browse/CDAP-3669
+    Map<ProgramType, Set<String>> programTypeToNames = new HashMap<>();
+    if (appSpec.getFlows() != null) {
+      programTypeToNames.put(ProgramType.FLOW, appSpec.getFlows().keySet());
+    }
+    if (appSpec.getMapReduce() != null) {
+      programTypeToNames.put(ProgramType.MAPREDUCE, appSpec.getMapReduce().keySet());
+    }
+    if (appSpec.getWorkflows() != null) {
+      programTypeToNames.put(ProgramType.WORKFLOW, appSpec.getWorkflows().keySet());
+    }
+    if (appSpec.getServices() != null) {
+      programTypeToNames.put(ProgramType.SERVICE, appSpec.getServices().keySet());
+    }
+    if (appSpec.getSpark() != null) {
+      programTypeToNames.put(ProgramType.SPARK, appSpec.getSpark().keySet());
+    }
+    if (appSpec.getWorkers() != null) {
+      programTypeToNames.put(ProgramType.WORKER, appSpec.getWorkers().keySet());
+    }
+
+    for (ProgramType programType : programTypeToNames.keySet()) {
+      Set<String> programNames = programTypeToNames.get(programType);
```

Comment:
@henry-cask , I think @poornachandra meant to say that using a loop on entryset instead of keyset might reduce another get that you have to do on line 739.